### PR TITLE
MRTK Oculus GO support

### PR DESCRIPTION
**Pull request by Alexees**
Overview
Since the Oculus Go is another device capable of using the MRTK and one that we will need in our upcoming switch from the very old MRTK to the new one, I integrate the Oculus Go into the kit.
It follows the same rules for controller and device manager creation as any other. Due to the fact it's the first Android platform though, a few adjustments to other scripts had to be made.

Changes
MRTK now supported Oculus Go

### CODE REVIEW

File: Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityEditorSettings.cs [hunk 1]
```diff
@@ -68,12 +68,14 @@ static MixedRealityEditorSettings()
                     message += "- Force Text Serialization\n";
                 }
 
+#if UNITY_WSA
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityEditorSettings.cs]: Why are we wrapping this?
Does this fix something we were not aware of?

File: Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs [hunk 2]
```diff
@@ -618,24 +618,24 @@ public bool RegisterService<T>(Type interfaceType, SupportedPlatforms supportedP
                 return false;
             }
 
-            if (interfaceType == null)
+#if !UNITY_EDITOR
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs]: These changes should be in a different pull request

File: Assets/MixedRealityToolkit.Providers/Oculus/GenericOculusAndroidController.cs [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+            new MixedRealityInteractionMapping(7, "Trackpad-Thumbstick Press", AxisType.Digital, DeviceInputType.TouchpadPress, KeyCode.JoystickButton8),
+            // HTC Vive Controller - Left Controller Menu Button (1)
+            // Oculus Touch Controller - Button.Three Press
+            // Valve Knuckles Controller - Left Controller Inner Face Button
+            // Windows Mixed Reality Controller - Left Menu Button
+            new MixedRealityInteractionMapping(8, "Unity Button Id 2", AxisType.Digital, DeviceInputType.ButtonPress, KeyCode.JoystickButton2),
+            // Oculus Touch Controller - Button.Four Press
+            // Valve Knuckles Controller - Left Controller Outer Face Button
+            new MixedRealityInteractionMapping(9, "Unity Button Id 3", AxisType.Digital, DeviceInputType.ButtonPress, KeyCode.JoystickButton3),
+            new MixedRealityInteractionMapping(10, "WMR Touchpad Touch", AxisType.Digital, DeviceInputType.TouchpadTouch, KeyCode.JoystickButton18),
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit.Providers/Oculus/GenericOculusAndroidController.cs]: Not sure if WMR is appropreate here.

File: Assets/MixedRealityToolkit.SDK/Profiles/DefaultMixedRealityRegisteredServiceProvidersProfile.asset [hunk 4]
```diff
@@ -19,7 +19,7 @@ MonoBehaviour:
         Microsoft.MixedReality.Toolkit
     componentName: Unity Joystick Manager
     priority: 10
-    runtimePlatform: -1
+    runtimePlatform: 31
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit.SDK/Profiles/DefaultMixedRealityRegisteredServiceProvidersProfile.asset]: Not sure this is a good change. Unity Joysticks should work on any platform

File: Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityEditorSettings.cs (hunk 1 above)
Comment (Alexees) [reply Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityEditorSettings.cs]: Since we're on Android, you don't want IL2CPP to be enforced. Code debugging is only possible on a C++ basis then. I was not sure which way to comment this, I started with !UNITY_ANDROID but was not sure.

File: Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs (hunk 2 above)
Comment (Alexees) [reply Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs]: That's true. How do I move them over?

File: Assets/MixedRealityToolkit.Providers/Oculus/GenericOculusAndroidController.cs (hunk 3 above)
Comment (Alexees) [reply Assets/MixedRealityToolkit.Providers/Oculus/GenericOculusAndroidController.cs]: This class  could very well be empty. I just copied another over so there was a generic version for any fallback to come. Since there is no fallback for Oculus GO I was not sure what a fallback could look like.

File: Assets/MixedRealityToolkit.SDK/Profiles/DefaultMixedRealityRegisteredServiceProvidersProfile.asset (hunk 4 above)
Comment (Alexees) [reply Assets/MixedRealityToolkit.SDK/Profiles/DefaultMixedRealityRegisteredServiceProvidersProfile.asset]: If not done this way the UnityJoystick will be picked, not the OculusAndroidDeviceManager I created for Android platform conciderations. I did not want to add the Oculus controller into the existing Manager so I added mine in 8th position, but to be reached, this has to be done.

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit.SDK/Profiles/DefaultMixedRealityRegisteredServiceProvidersProfile.asset]: Why create a controller that already handles this case?

Wouldn't it make more sense to have the generic unity joystick support this instead?

File: Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityEditorSettings.cs (hunk 1 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityEditorSettings.cs]: Same for UWP. Please revert this change.

File: Assets/MixedRealityToolkit.SDK/Profiles/DefaultMixedRealityRegisteredServiceProvidersProfile.asset (hunk 4 above)
Comment (Alexees) [reply Assets/MixedRealityToolkit.SDK/Profiles/DefaultMixedRealityRegisteredServiceProvidersProfile.asset]: I'm not sure. So far it only detects XBox controllers. the derived OpenVRDeviceManager detects all OpenVR controllers but only works for standalone. WindowsMixedRealityDeviceManager manages the WindowsMixedRealityController. This looks to me as if the UnityJoystickManager would be the best bet, but it does not make real sense either as every SDK (OpenVR, UWP ...kind of) has its manager.
If it follows a certain logic, why not .)

File: Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityEditorSettings.cs (hunk 1 above)
Comment (Alexees) [reply Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityEditorSettings.cs]: I don't understand. If I remove it, IL2CPP will be enforced, which is not good on Android but fine with UWP. Or did I missunderstood something?

File: Assets/MixedRealityToolkit.SDK/Profiles/DefaultMixedRealityRegisteredServiceProvidersProfile.asset (hunk 4 above)
Comment (Alexees) [reply Assets/MixedRealityToolkit.SDK/Profiles/DefaultMixedRealityRegisteredServiceProvidersProfile.asset]: what is really confusing is, to even get to the OpenVR device manager, the UnityJoystickManager gas to actively handle controllers with OpenVR in their names and revoke those

File: Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs (hunk 2 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs]: I'd suggest making a new branch from the latest development and copying over the changes.

Then reverting this file to what it was before.

File: Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityEditorSettings.cs (hunk 1 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityEditorSettings.cs]: It's not enforced. It asks you to update it. Although from what I understand IL2CPP for android is also the recommended setting as well, correct?

All of the settings suggested here are optional.

Comment (Railboy) [reply Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityEditorSettings.cs]: >Although from what I understand IL2CPP for android is also the recommended setting as well, correct?

We prefer to compile to .NET on Android. We'll often prototype / test iOS apps on Android because .NET helps with rapid iteration. We actually made the same exact change on our end because we were tired of seeing the warning. Worth opening another PR.

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityEditorSettings.cs]: > We prefer to compile to .NET on Android.

It's mono, not .NET

image

File: Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs (hunk 2 above)
Comment (Alexees) [reply Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs]: will do tomorrow

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs]: Thanks :)

File: Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityEditorSettings.cs (hunk 1 above)
Comment (Alexees) [reply Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityEditorSettings.cs]: I agree with Railboy. It's mandatory for UWP, most likely the better option on standalone, but not recommended on Android and therefore not worth the popup.
I will go with !UNITY_ANDROID then, ok?

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityEditorSettings.cs]: https://unity3d.com/learn/tutorials/topics/best-practices/il2cpp-mono

> You should use IL2CPP to both develop and ship your Project. If iteration times end up being too slow using IL2CPP, switch temporarily to the Mono Scripting Backend during development to improve iteration speed.

I can see the benefit for faster iteration times, but even Unity suggests IL2CPP for Android as well.
Isn't there a "don't show me this again" button on the popup to make that dialog disappear forever? I thought I added that.

**Comment by davidkline-ms** [comment]
Hi @Alexees! Thanks for the contribution! We are finalizing a plan where the community can release additional platform support as an extension.

Let me know if you have some time and we can discuss the plan in more detail.

I LOVE seeing this type of PR!

File: Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityEditorSettings.cs (hunk 1 above)
Comment (Alexees) [reply Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityEditorSettings.cs]: I check it out

**Comment by Alexees** [comment]
@davidkline-ms would love to, whenever you're ready during CET hours.

**Comment by Alexees** [comment]
@StephenHodgson I gave it another run, reduced and reverted to the minimum necessary for my setup

**Comment by Alexees** [comment]
@StephenHodgson I moved the change made to the MixedRealityToolkit.cs into its own PullRequest #3579

**Comment by davidkline-ms** [comment]
@Alexees, I am in the US Pacific time zone. I'll try to find something that overlaps at a comfortable time for us both.

I'm booked all this week, so will target sometime early next.

File: Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityEditorSettings.cs (hunk 1 above)
Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityEditorSettings.cs]: Agreed, our and Unity's recommendation is to use IL2CPP for all builds.  It's fine using mono for dev / testing, but IL2CPP for release.

Hence the startup prompt is optional and you can choose to ignore it as a dev

File: Assets/MixedRealityToolkit.Providers/Oculus/GenericOculusAndroidController.cs (hunk 3 above)
Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit.Providers/Oculus/GenericOculusAndroidController.cs]: Well, you should pre-fil it with appropriate defaults, the consumer has the full access to change this for their project.  But it should be clean and not retain any "copy" config / functions that are not specific to this device.

File: Assets/MixedRealityToolkit/StandardAssets/Textures/MotionController_left_black_scaled.png.meta [hunk 5]
```diff
@@ -3,7 +3,7 @@ guid: 03b731c99d54dfb4abde4d7d1b1b9199
 TextureImporter:
```
Comment (SimonDarksideJ) [inline Assets/MixedRealityToolkit/StandardAssets/Textures/MotionController_left_black_scaled.png.meta]: Can you please revert any .Meta changes not related to your change please.  Can't see any reason you are modifying the existing MotionController meta files.  Just double check the existing one's are not touched.

Comment (Alexees) [reply Assets/MixedRealityToolkit/StandardAssets/Textures/MotionController_left_black_scaled.png.meta]: I think I'm done with that.

Comment (Alexees) [reply Assets/MixedRealityToolkit/StandardAssets/Textures/MotionController_left_black_scaled.png.meta]: Referring to the Oculus GO SDK: You need to choose the Oculus Virtual Reality SDK in the XR Settings and have the Oculus (Android) package installed, that's true, but except for a specific configuration in the Manifest, that's about all you have to do to get it into the store I think.
The "Oculus Integration" Unity package from the store provides more stuff, but the last time I used it, from the input perspective, it wasn't nearly as complete as the MRTK (gaze and pointing, cursor, beam...)

**Comment by Alexees** [comment]
Sadly, the loading overhead of the toolkit on the Oculus Go is quite large (> 1s in frame time, but longer as observed in the Go).

Performance_Profile
1st_Frame_Init
2nd_Frame_Init
3rd_Frame_Init
4th_Frame_Init

**Comment by chrisfromwork** [comment]
Requesting some additional ms folks for visibility

**Comment by Alexees** [comment]
> @Alexees, I am in the US Pacific time zone. I'll try to find something that overlaps at a comfortable time for us both.
> 
> I'm booked all this week, so will target sometime early next.

@davidkline-ms are you still interested in a call? Surely I am.

**Comment by Railboy** [comment]
Now that the HL2 merge is complete I thought I'd ping this for visibility. Definitely interested in seeing this merged.

**Comment by davidkline-ms** [comment]
This should be released as an extension. I will work with @Alexees directly.

**Comment by Alexees** [comment]
@davidkline-ms Sounds great. Due to easter, I'll be back at work on April 23rd, maybe we could have a talk then?

**Comment by davidkline-ms** [comment]
@Alexees, sincere appologies for not having gotten back to this. I would very much like to chat (perhaps in email) about specifics of making this a community extension.

**Comment by Alexees** [comment]
@davidkline-ms sure we could do that. Since I don't have the full picture of the toolkit, maybe you could purpose a way to start this.
Probably by defining how an extension looks like. My mail address is as@hololux.de

**Comment by Alexees** [comment]
Following the suggestion of @davidkline-ms, this PR will find it's way a different way:
https://github.com/hololux/MRTKOculusGo

**Comment by StephenHodgson** [comment]
@Alexees please open a PR on xrtk

**Comment by Alexees** [comment]
@StephenHodgson is there a documentation about contribution repos? there seem to be more than obvious.

**Comment by chbecker-ms** [comment]
@davidkline-ms - Let's share with everyone how easy it is to build an extension for MRTK!

**Comment by StephenHodgson** [comment]
@Alexees I'm not sure I understand your question. I'm kind of disappointed that this PR went untended to for so long. Please open it directly in the xrtk fork as a PR I believe it belongs there as a core supported platform. We'll get it updated and ready to go in no time :)

**Comment by Alexees** [comment]
@StephenHodgson does core feature man it's not meant to be an extension m if so, you should talk to davidkline as he instructed me to handle it like one

**Comment by davidkline-ms** [comment]
> @davidkline-ms - Let's share with everyone how easy it is to build an extension for MRTK!

I will write something up early this week!

**Comment by StephenHodgson** [comment]
Seeing as the implementation you've generated is based off of the Unity input axes it can technically be part of the core stuff, although I think mrtk as made a decision to split them differently for what ever reason.

xrtk we try to split things logically by the low level input source. So if the data is coming from a specific platform/api that's always availible in Unity, then it's a core thing.

If it was only something availible specifically on Android, then it would be in the Android package/extension.

**Comment by Alexees** [comment]
@StephenHodgson I'll create the pull request in XRTK. What I meant with docu about contribution repos is that right now there seems to be about 3 repos that kind of are the newest. mrtk_development, 2019_rc1 and now this xrtk repo. I am constantly trying to use the most recent toolkit with fixes applied to it that I made. I fixed into mrtk_development, want to make use of Hololens 2 soon and want to use the Oculus Go implementation as soon as possible, but all of this goes into 3 different branches. How do these correlate?

**Comment by davidkline-ms** [comment]
@Alexees, xrtk is a separate product (managed by @StephenHodgson) from the MRTK (managed by Microsoft).

In the MRTK repository, the mrtk_development branch is where changes are checked in. When a release is being prepared, we create a stabilization branch (ex: rc1). Once stabilized, the code is merged into mrtk_release and is tagged.

I hope this helps clear up some of the branch questions.
